tests/test_dataclasses.py

```diff
@@ -1346,7 +1346,7 @@ class B:
 
     @pydantic.dataclasses.dataclass
     class Top:
-        sub: Union[A, B] = dataclasses.field(metadata=dict(discriminator='l'))
```

Comment:
I don't think we want to change most tests, just add some new tests specifically for the new behaviour.

Reply:
I've reverted the random test changes, updated existing tests more strategically where it seemed natural (i.e. they already used `from __future__ import annotations` or they were skipped for Python <= 3.9), and added new tests in various files (all called `test_eval_type_backport`) to test the different places where `eval_type_backport` is called.